### PR TITLE
 [14.0][ADD] sale_timesheet_order_line_no_update

### PR DESCRIPTION
**Pull request by ilyasProgrammer**
This module allows to set which Sale order item to use for new timesheets in a specific task, without updating the related Sale order item for all the existing timesheets in that task. This value can be set in field "Default sale order item" in the task itself.

In project > invoicing, updating field "Default Sales Order Item" updates new field "Default sale order item" in all tasks.

It also allows to select in both project and tasks a Sale order item from a different SO than the one set in project configuration.

### CODE REVIEW

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

File: sale_timesheet_order_line_no_update/models/account_analytic_line.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        skip_ids = self.filtered(lambda t: t.task_id and t.task_id.new_sale_line_id)
+        self -= skip_ids
+        if self:
+            super(
+                AccountAnalyticLine, self.filtered(lambda t: not t.is_so_line_edited)
+            )._compute_so_line()
+
+    @api.model
+    def _timesheet_determine_sale_line(self, task, employee, project):
+        if task and task.allow_billable and task.sale_line_id:
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/account_analytic_line.py]: The best solution would be joining conditions into one:
```
if task and task.allow_billable and task.sale_line_id and task.pricing_type == "fixed_rate" and task.new_sale_line_id:
    ...
```

File: sale_timesheet_order_line_no_update/models/account_analytic_line.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            if task.pricing_type == "fixed_rate":
+                if task.new_sale_line_id:
+                    return task.new_sale_line_id
+        return super(AccountAnalyticLine, self)._timesheet_determine_sale_line(
+            task, employee, project
+        )
+
+    @api.onchange("employee_id")
+    def _onchange_task_id_employee_id(self):
+        if self.project_id and self.task_id.allow_billable:  # timesheet only
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/account_analytic_line.py:37]: The best solution would be joining conditions into one:
```
if self.project_id and self.task_id.allow_billable and (self.task_id.bill_type == "customer_task" or self.task_id.pricing_type == "fixed_rate"):
```

File: sale_timesheet_order_line_no_update/models/project.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        readonly=True,
+        store=False,
+    )
+
+    @api.depends("project_id", "project_id.select_all_project_sale_items")
+    def _compute_new_sale_line_id_domain(self):
+        for rec in self:
+            if not self.project_id.select_all_project_sale_items:
+                rec.new_sale_line_id_domain = json.dumps(
+                    [
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/project.py]: These domains are duplicated in function _compute_sale_line_id_domain project.project model.
Could you fix it?

File: sale_timesheet_order_line_no_update/models/project.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                        ("state", "in", ["sale", "done"]),
+                        "|",
+                        ("company_id", "=", False),
+                        ("company_id", "=", rec.company_id.id),
+                    ]
+                )
+
+    @api.onchange("sale_line_id")
+    def _onchange_sale_line_id(self):
+        task_ids = self.env["project.task"].search([("project_id", "in", self.ids)])
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/project.py]: You can use related tasks for projects:
`task_ids = self.mapped('tasks')`

File: sale_timesheet_order_line_no_update/models/project.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        compute="_compute_sale_line_id_domain",
+        readonly=True,
+        store=False,
+    )
+
+    @api.depends("select_all_project_sale_items")
+    def _compute_sale_line_id_domain(self):
+        for rec in self:
+            if not self.select_all_project_sale_items:
+                rec.sale_line_id_domain = json.dumps(
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/project.py]: You can prepare domain and then write it in 'sale_line_id_domain' field.

**Review by geomer198 — COMMENTED** [review]
Please add tests and docstring for all functions.

File: sale_timesheet_order_line_no_update/models/project.py (hunk 5 above)
Comment (ilyasProgrammer) [reply sale_timesheet_order_line_no_update/models/project.py]: Can you write down what you mean ?

Comment (geomer198) [reply sale_timesheet_order_line_no_update/models/project.py]: For example:
```
def _compute_sale_line_id_domain(self):
    for rec in self:
         field, value = "order_partner_id", rec.partner_id.id  if rec.select_all_project_sale_items else "order_id", rec.sale_order_id.id
        rec.sale_line_id_domain = [
                        ("is_service", "=", True),
                        ("is_expense", "=", False),
                        (field, "=", value),
                        ("state", "in", ["sale", "done"]),
                        "|",
                        ("company_id", "=", False),
                        ("company_id", "=", rec.company_id.id),
                    ]
```

File: sale_timesheet_order_line_no_update/models/project.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    sale_line_id_domain = fields.Char(
+        compute="_compute_sale_line_id_domain",
+        readonly=True,
+        store=False,
+    )
+
+    @api.depends("select_all_project_sale_items")
+    def _compute_sale_line_id_domain(self):
+        for rec in self:
+            if not self.select_all_project_sale_items:
```
Comment (geomer198) [inline sale_timesheet_order_line_no_update/models/project.py]: Why do you use `self.select_all_project_sale_items`, may be you should use `rec.select_all_project_sale_items` instead?

**Review by geomer198 — APPROVED** [review]
LGTM!
Code Review ok!

**Comment by francesco-ooops** [comment]
@OCA/human-resources-maintainers does this solution look ok for you?

**Comment by francesco-ooops** [comment]
@ilyasProgrammer please squash and remove version from commit name

**Comment by francesco-ooops** [comment]
@OCA/human-resources-maintainers tests added and tag "ready to merge" is here, can you take a look? thank you

**Comment by francesco-ooops** [comment]
@OCA/human-resources-maintainers merge?

File: sale_timesheet_order_line_no_update/models/account_analytic_line.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+from odoo import api, models
```
Comment (Saran440) [inline sale_timesheet_order_line_no_update/models/account_analytic_line.py:3]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).

from odoo import api, models
```

File: sale_timesheet_order_line_no_update/models/account_analytic_line.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    def _timesheet_determine_sale_line(self, task, employee, project):
+        if (
+            task
+            and task.allow_billable
+            and task.sale_line_id
+            and task.pricing_type == "fixed_rate"
+            and task.new_sale_line_id
+        ):
+            return task.new_sale_line_id
+        return super(AccountAnalyticLine, self)._timesheet_determine_sale_line(
```
Comment (Saran440) [inline sale_timesheet_order_line_no_update/models/account_analytic_line.py]:
```suggestion
        return super()._timesheet_determine_sale_line(
```

File: sale_timesheet_order_line_no_update/models/project.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+from odoo import api, fields, models
```
Comment (Saran440) [inline sale_timesheet_order_line_no_update/models/project.py:3]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).

from odoo import api, fields, models
```

File: sale_timesheet_order_line_no_update/models/res_config_settings.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+from odoo import fields, models
```
Comment (Saran440) [inline sale_timesheet_order_line_no_update/models/res_config_settings.py:3]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).

from odoo import fields, models
```

File: sale_timesheet_order_line_no_update/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (Saran440) [inline sale_timesheet_order_line_no_update/__manifest__.py:3]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).

{
```

**Review by Saran440 — COMMENTED** [review]
Small fix.

**Comment by ilyasProgrammer** [comment]
updated

**Comment by francesco-ooops** [comment]
@Saran440 can you take another look? Thanks!

**Pull request merged**